Ethernet bridge forward decision and IP layer communication support

yangbolu1991 opened the pull request:
Ethernet bridge currently supported very sample forwarding for all input packets.
We will have more features development based on virtual bridge, like IP communication, fdb, vlan ...

So, this PR did below work.
- Moved bridge code to folder.
- Handled bridge input with added forward decision.
- Supported virtual bridge RX and IP protocols.
- Added bridge sample with DHCP and added i.mxrt1180 board as an example.

In the sample, ports were bridged and virtual bridge interface was with DHCP client enabled to get IP.
For i.MXRT 1180, 5 ports were bridged including 4 DSA ports and 1 normal Ethernet port. See the log below.

```
[00:00:00.052,000] <inf> phy_mii: PHY (3) ID 1CC816
[00:00:00.103,000] <inf> phy_mii: PHY (2) ID 1CC816
[00:00:00.140,000] <err> net_dsa_tag: Failed to connect tag protocol
[00:00:00.143,000] <inf> dsa_netc: DSA user port 0 Link down
[00:00:00.144,000] <inf> dsa_netc: DSA user port 1 Link down
[00:00:00.144,000] <inf> dsa_netc: DSA user port 2 Link down
[00:00:00.144,000] <inf> dsa_netc: DSA user port 3 Link down
[00:00:00.144,000] <inf> nxp_imx_eth_psi: ENETC0 MAC: 00:00:00:01:02:00
[00:00:00.144,000] <inf> nxp_imx_eth_psi: ENETC0 Link down
[00:00:00.144,000] <inf> nxp_imx_eth_psi: ENETC1 MAC: 00:00:00:01:03:00
*** Booting Zephyr OS build v4.3.0-339-g698792a29a43 ***
[00:00:00.145,000] <inf> net_bridge_sample: Find bridge iface 1.
[00:00:00.145,000] <inf> net_bridge_sample: Find iface 2. Add into bridge.
[00:00:00.146,000] <inf> net_eth_bridge: Bridge 1 is setup
[00:00:00.146,000] <inf> net_bridge_sample: Find iface 3. Add into bridge.
[00:00:00.146,000] <inf> net_eth_bridge: Bridge 1 is setup
[00:00:00.146,000] <inf> net_bridge_sample: Find iface 4. Add into bridge.
[00:00:00.146,000] <inf> net_eth_bridge: Bridge 1 is setup
[00:00:00.146,000] <inf> net_bridge_sample: Find iface 5. Add into bridge.
[00:00:00.146,000] <inf> net_bridge_sample: Find iface 6. Add into bridge.
[00:00:00.146,000] <inf> net_eth_bridge: Bridge 1 is setup
[00:00:00.146,000] <inf> net_bridge_sample: Find iface 7. Add into bridge.
[00:00:00.146,000] <inf> net_eth_bridge: Bridge 1 is active
uart:~$
uart:~$
uart:~$
[00:00:03.238,000] <inf> phy_mii: PHY (2) Link speed 100 Mb, full duplex
[00:00:03.239,000] <inf> dsa_netc: DSA user port 0 Link up
[00:00:03.615,000] <inf> phy_rt_rtl8211f: PHY 4 is up
[00:00:03.615,000] <inf> phy_rt_rtl8211f: PHY (4) Link speed 1000 Mb, full duplex
[00:00:03.615,000] <inf> dsa_netc: DSA user port 2 Link up
uart:~$
uart:~$
[00:00:06.642,000] <inf> net_dhcpv4: Received: 192.168.0.102
uart:~$
uart:~$
uart:~$ net ping 192.168.0.10
PING 192.168.0.10
28 bytes from 192.168.0.10 to 192.168.0.102: icmp_seq=1 ttl=64 time=5 ms
28 bytes from 192.168.0.10 to 192.168.0.102: icmp_seq=2 ttl=64 time=2 ms
28 bytes from 192.168.0.10 to 192.168.0.102: icmp_seq=3 ttl=64 time=2 ms
uart:~$
uart:~$
uart:~$ net bridge
Bridge Status   Config   Interfaces
1      up       ok       2 3 4 5 7
uart:~$
uart:~$ net iface 1
Default interface: 1


Interface bridge0 (0x1400b638) (Virtual) [1]
==================================
Virtual name : <enabled>
No attached network interface.
Link addr : EF:75:EE:44:74:19
MTU       : 1500
Flags     : NO_AUTO_START,IPv4,IPv6
Device    : BRIDGE_0 (0x38035a58)
Status    : oper=UP, admin=UP, carrier=ON
Promiscuous mode : disabled
IPv6 unicast addresses (max 2):
        fe80::ef75:eeff:fe44:7419 autoconf preferred infinite
IPv6 multicast addresses (max 3):
        <none>
IPv6 prefixes (max 2):
        <none>
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 29063
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        192.168.0.102/255.255.0.0 DHCP preferred
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 192.168.0.2
DHCPv4 lease time : 3600
DHCPv4 renew time : 1800
DHCPv4 server     : 192.168.0.10
DHCPv4 requested  : 192.168.0.102
DHCPv4 state      : bound
DHCPv4 attempts   : 1
DHCPv4 state      : bound
uart:~$

```

Thanks a lot.

yangbolu1991 committed:
net: bridge: handle RX in bridge_input.c

Current bridge function is very simple which does forwarding for all
packets. There will be more and more features of bridge RX handling.

Let's move bridge RX handling to bridge_input.c for next IPv4/IPv6
protocols support for virtual bridge interface.

And currently there is no path to call bridge_iface_recv function.
Let's rework it for proper function returning NET_CONTINUE.

Also fixed another issue of link local address checking. Because
net_linkaddr structure changed. So fixed

is_link_local_addr((struct net_eth_addr *)net_pkt_lladdr_dst(pkt))

to

is_link_local_addr((struct net_eth_addr *)(net_pkt_lladdr_dst(pkt)->addr))

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: ethernet: support bridge RX

Added support for bridge RX. We were going to support protocols
on bridge.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: bridge: add forward decision support

Added forward decision support including for broadcast,
multicast and local address packets.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: bridge: enable IPv4/IPv6 support

Ethernet bridge performs L2 forwarding between bridged interfaces.
But for the virtual bridge interface, we could make it perform as
normal Ethernet interface with IP address to use.

This is useful allowing applications to communicate with network,
like LAN router or switch with a host IP.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
net: net_if: use debug level message for iface down

Dropping packet is normal operation when iface is down.
So, we don't have to use warning here for much noise.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: add Ethernet bridge sample

Added Ethernet bridge sample and enable mimxrt1180_evk_mimxrt1189_cm33
board as an example. On mimxrt1180_evk_mimxrt1189_cm33, 5 ports are
bridged with DHCPv4 enabled to test, including 4 DSA ports and
1 normal Ethernet port.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

jukkar reviewed (commented):
Thanks for this. Couple of comments to the sample app. I think we have now multiple samples related to ethernet, so it might make sense to create `samples/net/ethernet` and place this bridge sample and existing gptp, lldp and vlan samples there.
Other thing about the bridge sample, it needs to support native_sim application somehow, otherwise testing / trying bridge related things is limited to rather small audience.

yangbolu1991 committed:
net: dsa: use snprintk instead of snprintf

Used snprintk instead of snprintf as snprintk is compatible
across platforms.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: move Ethernet samples to folder

Moved current Ethernet samples to folder ethernet,
as there will be more samples added in the future.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
MAINTAINERS: update vlan and dsa samples path

Updated vlan and dsa samples path as they were moved.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
> Thanks for this. Couple of comments to the sample app. I think we have now multiple samples related to ethernet, so it might make sense to create `samples/net/ethernet` and place this bridge sample and existing gptp, lldp and vlan samples there. Other thing about the bridge sample, it needs to support native_sim application somehow, otherwise testing / trying bridge related things is limited to rather small audience.

Thanks @jukkar 

Updated to v2. Changes include.
- Created ethernet folder in samples and moved related lldp, vlan, gptp and dsa to it.
- Added native_sim platform support in bridge. (Not sure if it's enough. Or you want me to write some more host shell script for testing?)

Thank you very much.

jukkar commented on subsys/net/l2/ethernet/bridge/bridge.c:
```diff
@@ -255,8 +255,8 @@ static void bridge_iface_init(struct net_if *iface)
-	net_if_flag_clear(iface, NET_IF_IPV4);
-	net_if_flag_clear(iface, NET_IF_IPV6);
+	net_if_flag_set(iface, NET_IF_IPV4);
+	net_if_flag_set(iface, NET_IF_IPV6);
```
Why do we need to do this as bridging is done in L2?

jukkar replied:
Fair enough. I would suggest adding to the commit message the reasoning too, new it states the obvious thing which we can see from the changes but does not answer the question "why this change is done".

yangbolu1991 commented:
Updated to v3. Changes include,
- Rebased to address conflict.
- Fixed doc issues.
- Added script and readme for verify bridging with native_sim.

Thanks a lot.

maass-hamburg commented on subsys/net/l2/ethernet/bridge/bridge_input.c:
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2025 NXP
```
use https://docs.zephyrproject.org/latest/contribute/guidelines.html#copyright-and-license-notices (`SPDX-FileCopyrightText:`)

yangbolu1991 replied:
Thanks. Will fix for all new added copyright.

yangbolu1991 commented:
Updated to v4. Changes include,
- No code changes. Just improved commit message per @jukkar  suggestion.

Thanks.

yangbolu1991 committed:
net: bridge: move bridge files to folder

Moved bridge files to folder to manage. More bridge features
and files will be added in the future.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented on include/zephyr/net/ethernet_bridge.h:
```diff
@@ -136,6 +138,59 @@ typedef void (*eth_bridge_cb_t)(struct eth_bridge_iface_context *br, void *user_
+	return false;
+#else
+	return false;
+#endif
```
Thanks. Will do that.

yangbolu1991 commented:
Updated to v5. Changes include,
- Rebased to address conflict.
- Code improvement and copyright format fix-up per @maass-hamburg  comments.

Thanks.

yangbolu1991 commented on samples/net/ethernet/bridge/src/main.c:
```diff
@@ -0,0 +1,100 @@
+#endif
+};
+
+struct ud user_data = {0};
```
Thanks @jukkar 

Updated to v6. Changes include,
- Renamed global variable to avoid variable shadowing warning in sonarcloud.

yangbolu1991 commented:
> @yangbolu1991 please also take a look at the found issues from sonarqubecloud

Since I moved all ethernet samples to folder, the orignal code issues exposed now.
Do you want me to append some more patches to fix them in this PR?

Thanks

maass-hamburg commented:
> > @yangbolu1991 please also take a look at the found issues from sonarqubecloud
> 
> Since I moved all ethernet samples to folder, the orignal code issues exposed now. Do you want me to append some more patches to fix them in this PR?
> 
> Thanks

If it is not to much for you, fixing them now would be nice. Otherwise it would probably be forgotten again. (non blocking)

yangbolu1991 committed:
samples: net: dsa: improve code quality

Improved code quality with below changes.
- Avoided declaration shadowing variable in the global scope.
- Avoided missing default in switch block.
- Dropped unused variable.
- Reduced the level of if statements properly.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: gptp: add default case for status switch statement

Added default case for status switch statement.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

jukkar commented on samples/net/ethernet/gptp/docker-test.sh:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
-if [ -z "$RUNNING_FROM_MAIN_SCRIPT" ]; then
+if [[ -z "$RUNNING_FROM_MAIN_SCRIPT" ]]; then
```
What is the idea here, why do we need to do these changes in this file?

jukkar reviewed (changes requested):
About the new commits you added to fix the sonarcloud warnings, I would prefer that the commit message would give more information what is actually being fixed. Some of the changes do not make much sense unless more information is added to the commit message (for example you could add the warning text etc)
